File: distributed-job-manager/src/test/java/ru/fix/distributed/job/manager/strategy/RendezvousHashAssignmentStrategyTest.java
package ru.fix.distributed.job.manager.strategy;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.fix.distributed.job.manager.model.AssignmentState;
import ru.fix.distributed.job.manager.model.JobId;
import ru.fix.distributed.job.manager.model.WorkItem;
import ru.fix.distributed.job.manager.model.WorkerId;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static ru.fix.distributed.job.manager.strategy.AssignmentStrategyUtils.*;

class RendezvousHashAssignmentStrategyTest {
    private RendezvousHashAssignmentStrategy rendezvous;

    @BeforeEach
    void setUp() {
        rendezvous = new RendezvousHashAssignmentStrategy();
    }

    @Test
    void reassignAndBalanceWhenOnlyOneWorkerHasJobs() {
        AssignmentState available = new AssignmentState();
        AssignmentState previous = new AssignmentState();

        available.addWorkItem(new WorkerId("worker-0"), new WorkItem("work-item-0", new JobId("job-0")));
        available.addWorkItem(new WorkerId("worker-0"), new WorkItem("work-item-1", new JobId("job-0")));
        available.addWorkItem(new WorkerId("worker-0"), new WorkItem("work-item-2", new JobId("job-0")));
        available.addWorkItem(new WorkerId("worker-0"), new WorkItem("work-item-0", new JobId("job-1")));
        available.addWorkItem(new WorkerId("worker-0"), new WorkItem("work-item-1", new JobId("job-1")));
        available.put(new WorkerId("worker-1"), new HashSet<>());
        available.put(new WorkerId("worker-2"), new HashSet<>());

        previous.put(new WorkerId("worker-0"), new HashSet<>());
        previous.put(new WorkerId("worker-1"), new HashSet<>());
        previous.put(new WorkerId("worker-2"), new HashSet<>());

        AssignmentState newAssignment = rendezvous.reassignAndBalance(
                generateAvailability(available),
                previous,
                new AssignmentState(),
                generateItemsToAssign(available)
        );

        assertEquals(available.globalPoolSize(), newAssignment.globalPoolSize());
    }

    @Test
    void reassignAndBalanceWhenSomeWorkersHasJobs() {
        AssignmentState available = new AssignmentState();
        AssignmentState previous = new AssignmentState();

        addWorkerWithItems(available, "worker-0", 3, 3);
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-1", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-2", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-0", new JobId("job-3")));
        available.put(new WorkerId("worker-2"), new HashSet<>());

        previous.put(new WorkerId("worker-0"), new HashSet<>());
        previous.put(new WorkerId("worker-1"), new HashSet<>());
        previous.put(new WorkerId("worker-2"), new HashSet<>());

        assertFalse(available.isBalanced());

        AssignmentState newAssignment = rendezvous.reassignAndBalance(
                generateAvailability(available),
                previous,
                new AssignmentState(),
                generateItemsToAssign(available)
        );

        assertEquals(available.globalPoolSize(), newAssignment.globalPoolSize());
    }

    @Test
    void reassignAndBalanceIfWorkerNotAvailable() {
        AssignmentState available = new AssignmentState();
        AssignmentState previous = new AssignmentState();

        addWorkerWithItems(available, "worker-0", 3, 1);

        previous.addWorkItems(new WorkerId("worker-1"), Set.of(
                new WorkItem("work-item-2", new JobId("job-3")),
                new WorkItem("work-item-0", new JobId("job-3"))
        ));
        previous.addWorkItems(new WorkerId("worker-0"), Set.of(
                new WorkItem("work-item-1", new JobId("job-3")),
                new WorkItem("work-item-0", new JobId("job-0"))
        ));

        AssignmentState newAssignment = rendezvous.reassignAndBalance(
                generateAvailability(available),
                previous,
                new AssignmentState(),
                generateItemsToAssign(available)
        );

        assertEquals(available.globalPoolSize(), newAssignment.globalPoolSize());
    }

    @Test
    void reassignAndBalanceIfNewWorkersAdded() {
        AssignmentState available = new AssignmentState();
        AssignmentState previous = new AssignmentState();

        addWorkerWithItems(available, "worker-0", 3, 1);
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-1", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-2", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-0", new JobId("job-3")));
        available.addWorkItems(new WorkerId("worker-2"), Collections.emptySet());
        available.addWorkItems(new WorkerId("worker-3"), Collections.emptySet());

        previous.addWorkItems(new WorkerId("worker-0"), Set.of(
                new WorkItem("work-item-1", new JobId("job-3")),
                new WorkItem("work-item-0", new JobId("job-0"))
        ));
        previous.addWorkItems(new WorkerId("worker-1"), Set.of(
                new WorkItem("work-item-0", new JobId("job-3"))
        ));

        AssignmentState newAssignment = rendezvous.reassignAndBalance(
                generateAvailability(available),
                previous,
                new AssignmentState(),
                generateItemsToAssign(available)
        );

        assertEquals(available.globalPoolSize(), newAssignment.globalPoolSize());
    }

    @Test
    void reassignAndBalanceIfWorkerNotAvailableAndNewWorkerAdded() {
        AssignmentState available = new AssignmentState();
        AssignmentState previous = new AssignmentState();

        addWorkerWithItems(available, "worker-0", 3, 1);
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-1", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-2", new JobId("job-3")));
        available.addWorkItem(new WorkerId("worker-1"), new WorkItem("work-item-0", new JobId("job-3")));

        // Previous state contains worker-2 instead of worker-1.
        // It's emulate case, when worker-1 is not available, and worker-2 connected
        previous.addWorkItems(new WorkerId("worker-0"), Set.of(
                new WorkItem("work-item-1", new JobId("job-3")),
                new WorkItem("work-item-0", new JobId("job-0"))
        ));
        previous.addWorkItems(new WorkerId("worker-2"), Set.of(
                new WorkItem("work-item-0", new JobId("job-3")),
                new WorkItem("work-item-2", new JobId("job-0")),
                new WorkItem("work-item-1", new JobId("job-0")),
                new WorkItem("work-item-2", new JobId("job-3"))
        ));

        AssignmentState newAssignment = rendezvous.reassignAndBalance(
                generateAvailability(available),
                previous,
                new AssignmentState(),
                generateItemsToAssign(available)
        );

        assertEquals(previous.globalPoolSize(), newAssignment.globalPoolSize());
    }
}

File: distributed-job-manager/src/test/kotlin/ru/fix/distributed/job/manager/strategy/RendezvousHashAssignmentStrategyTest.kt
package ru.fix.distributed.job.manager.strategy

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import ru.fix.distributed.job.manager.JobId
import ru.fix.distributed.job.manager.model.AssignmentState
import ru.fix.distributed.job.manager.model.WorkerId

internal class RendezvousHashAssignmentStrategyTest {
    private lateinit var rendezvous: RendezvousHashAssignmentStrategy

    companion object {
        val logger: Logger = LoggerFactory.getLogger(RendezvousHashAssignmentStrategyTest::class.java)
    }

    @BeforeEach
    fun setUp() {
        rendezvous = RendezvousHashAssignmentStrategy()
    }

    private val workPool: JobScope.() -> Unit = {
        "job-0"(
            "work-item-0",
            "work-item-1",
            "work-item-2",
            "work-item-3",
            "work-item-4",
            "work-item-5"
        )
    }
    private val workPool1: JobScope.() -> Unit = {
        "job-1"(
            "work-item-0",
            "work-item-1",
            "work-item-2"
        )
    }

    @Test
    fun `assign items when previous state empty`() {
        val available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool)
            "worker-2"(workPool)
        }
        val previous = assignmentState {
            "worker-0"{}
            "worker-1"{}
            "worker-2"{}
        }
        val newAssignment = calculateNewAssignment(available, previous)
        assertEquals(6, newAssignment.globalPoolSize())
    }

    @Test
    fun `reassign items when new worker added`() {
        var available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool1)
            "worker-2"(workPool)
        }
        var previous = assignmentState {
            "worker-0"{}
            "worker-1"{}
            "worker-2"{}
        }

        previous = calculateNewAssignment(available, previous)
        available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool1)
            "worker-2"(workPool)
            "worker-3"(workPool)
        }

        val newAssignment = calculateNewAssignment(available, previous)
        assertEquals(previous.globalPoolSize(), newAssignment.globalPoolSize())

        val workPoolSizeJob1 = previous.localPoolSize(JobId("job-1"))
        val workPoolSizeJob1OnWorker1 = newAssignment.getWorkItems(WorkerId("worker-1"), JobId("job-1")).size
        assertEquals(workPoolSizeJob1, newAssignment.localPoolSize(JobId("job-1")))
        assertEquals(workPoolSizeJob1, workPoolSizeJob1OnWorker1)
    }

    @Test
    fun reassignAndBalanceIfWorkerNotAvailable() {
        var available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool1)
        }
        var previous = assignmentState {
            "worker-0"{}
            "worker-1"{}
        }
        previous = calculateNewAssignment(available, previous)
        available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool1)
            "worker-2"(workPool)
        }

        val newAssignment = calculateNewAssignment(available, previous)
        assertEquals(previous.globalPoolSize(), newAssignment.globalPoolSize())
    }

    @Test
    fun reassignAndBalanceIfNewWorkersAdded() {
        var available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool)
            "worker-2"(workPool)
        }
        var previous = assignmentState {
            "worker-0"{}
            "worker-1"{}
            "worker-2"{}
        }

        previous = calculateNewAssignment(available, previous)
        available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool)
            "worker-2"(workPool)
            "worker-3"(workPool)
        }
        println(previous)

        val newAssignment = calculateNewAssignment(available, previous)
        assertEquals(previous.globalPoolSize(), newAssignment.globalPoolSize())
    }

    @Test
    fun reassignAndBalanceIfWorkerNotAvailableAndNewWorkerAdded() {
        var available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool)
            "worker-2"(workPool)
        }
        var previous = assignmentState {
            "worker-0"{}
            "worker-1"{}
            "worker-2"{}
        }

        previous = calculateNewAssignment(available, previous)
        available = assignmentState {
            "worker-0"(workPool)
            "worker-1"(workPool)
            "worker-3"(workPool)
        }
        val newAssignment = calculateNewAssignment(available, previous)
        assertEquals(previous.globalPoolSize(), newAssignment.globalPoolSize())
    }

    private fun calculateNewAssignment(
        available: AssignmentState,
        previous: AssignmentState
    ): AssignmentState {
        val availability = generateAvailability(available)
        val itemsToAssign = generateItemsToAssign(available)

        logger.info(
            Report(
                availability,
                itemsToAssign,
                previous
            ).toString()
        )

        val newState = AssignmentState()
        rendezvous.reassignAndBalance(
            availability,
            previous,
            newState,
            itemsToAssign
        )
        logger.info(Report(newAssignment = newState).toString())
        return newState
    }
}